### PR TITLE
Add improved pac4j-kerberos

### PR DESCRIPTION
**Pull request by vidma**
- [x] Rebase https://github.com/pac4j/pac4j/pull/595/ over master, fixing changes to the APIs
  * DirectClientV2, io.Resource directly from springframework etc
  * update `since 2.1.0`
- [x] Cleanup/simplify code, indentation, etc
  * Use regular HttpAction instead of a custom DeferredHttpAction
  * Fix bug when absolute keytab path was lost (using only filename)
- [x] Add integration/unit test with Kerby
  * mocked web-context
  * real Kerberos keytabs and real SPNEGO tickets
- [x] Add/split into `IndirectKerberosClient` and `DirectKerberosClient`, [discussion here](https://github.com/pac4j/pac4j/pull/919#discussion_r118471748)
- [x] Update/add tests for Direct/Indirect
- [ ] ~~Direct client with FORCE_AUTH_IF_NO_CREDENTIALS option (postponed, can be implemented easily without tweeking core code)~~.
- [x] Add docs
  * basically as done in `setupKerberosClient`, could also describe some common pitfalls:
    - what should be the service pricipal, i.e. `HTTP/full-qualified-domain-name@DOMAIN`
    - Java Cryptography Extension (JCE)
- [x] test again with `play-pac4` example with a real Kerberos setup. works fine.

fixes https://github.com/pac4j/pac4j/issues/548
@leleuj   P.S. my actual changes are only in the last two commits.

### CODE REVIEW

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTest.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * This is a test with real kerberos ticket validation:
+ * - spins up a mini/lightweight Kerberos server (using Apache Kerby)
+ * - generates the keytabs for service and client
+ * - generates a real service ticket, and passes it to mock context
+ * - checks that the ticket is correctly validated, and yields a correct client ID
+ *
+ * @author Vidmantas Zemleris, at Kensu.io
+ */
+public class KerberosClientKerbyTest implements TestsConstants {
```
Comment (vidma) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTest.java]: it's somewhere between integration and unit test.
likely I should make it run together with the integration tests...

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/profile/KerberosProfile.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+/**
+ * Represents a user profile based on a Kerberos authentication.
+ *
+ * @author Garry Boyce
+ * @since 2.1.0
+ */
+public class KerberosProfile extends CommonProfile {
+
+    private static final long serialVersionUID = -1388563485891552197L;
+    public GSSContext gssContext = null;
```
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/profile/KerberosProfile.java]: added optional `gssContext` for more advanced usage, if any.
if `SunJaasKerberosTicketValidator.setHoldOnToGSSContext(true)` was called, when this would allow subsequent encrypt/decrypt operations.  see https://github.com/pac4j/pac4j/pull/919/files#diff-c6ef70675c7cb3b2df3dde625bac09ecR124

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    protected void clientInit(final WebContext context) {
+        setCredentialsExtractor(new KerberosExtractor(getName()));
+        // FIXME: is this needed
+        assertAuthenticatorTypes(KerberosAuthenticator.class);
+    }
+
+    // FIXME: this was copied from earlier version
+    protected void assertAuthenticatorTypes(final Class<? extends Authenticator>... classes) {
```
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java]: this fn was part of pac4j `DirectClient*`, but was removed. do you remember why? @leleuj

Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java]: This was a check on the `Authenticator` type, but it was definitely too complicated and a bit too stringent.

During the development of the 2.0.0 version, I finally decided to remove it. You can delete it also.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     * @param ticketValidator The ticket validator used to validate the Kerberos ticket.
+     * @since 2.1.0
+     */
+    public KerberosAuthenticator(KerberosTicketValidator ticketValidator) {
+        this.ticketValidator = ticketValidator;
+    }
+
+    @Override
+    public void validate(KerberosCredentials credentials, WebContext context) throws CredentialsException {
+        logger.trace("Try to validate Kerberos Token:" + credentials.getKerberosTicketAsString());
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java]: The call to the `init()` method is missing here.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            subject = KerberosProfile.class.getSimpleName() + UserProfile.SEPARATOR + subject;
+        }
+        logger.debug("Succesfully validated, updated subject: " + subject);
+    }
+
+    @Override
+    protected void internalInit(WebContext context) {
+        CommonHelper.assertNotNull("ticketValidator", this.ticketValidator);
+        if (this.ticketValidator instanceof InitializableObject) {
+            ((InitializableObject) this.ticketValidator).init();
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java]: This is the way I used to initialize objects, but I tend to do things differently right now: initialization must be handled by the component which is initializable, not by another one using it; otherwise, initialization may be forgotten.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/SunJaasKerberosTicketValidator.java [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        try {
+            return Subject.doAs(this.serviceSubject, new KerberosValidateAction(token));
+        } catch (PrivilegedActionException e) {
+            throw new BadCredentialsException("Kerberos validation not successful", e);
+        }
+    }
+
+    @Override
+    protected void internalInit() {
+        try {
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/SunJaasKerberosTicketValidator.java]: We should call the `init()` method first, instead of letting the `KerberosAuthenticator` do that for us.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/SunJaasKerberosTicketValidator.java [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    private String servicePrincipal;
+    private Resource keyTabLocation;
+    private Subject serviceSubject;
+    private boolean holdOnToGSSContext;
+    private boolean debug = false;
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Override
+    public KerberosTicketValidation validateTicket(byte[] token) throws BadCredentialsException {
+        try {
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/SunJaasKerberosTicketValidator.java]: We should call the `init()` method first, instead of letting the KerberosAuthenticator do that for us.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+        String subject = ticketValidation.username();
+        logger.debug("Succesfully validated " + subject);
+
+        KerberosProfile profile = new KerberosProfile();
+        profile.setId(subject);
+        profile.gssContext = ticketValidation.getGssContext();
+        credentials.setUserProfile(profile);
+
+        // FIXME: which one do we prefer to use a profile ID? `KerberosProfile#garry vs garry` ?
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java]: Can you elaborate a little bit on this?

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+    public KerberosExtractor(final String clientName) {
+        this.clientName = clientName;
+    }
+
+    @Override
+    public KerberosCredentials extract(WebContext context) throws CredentialsException, HttpAction {
+        final String header = context.getRequestHeader(HttpConstants.AUTHORIZATION_HEADER);
+        if (header == null) {
+            throw HttpAction.unauthorizedNegotiate("Kerberos Header not found", context);
```
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java]: so, @leleuj , looking at other extractors/clients,  I guess this exception throwing is not the responsibility of `KerberosExtractor` and should be moved to `Kerberos(In)?DirectClient`.

there is one semantic nuance with SPNEGO/Kerberos - the browser (e.g. Firefox) do NOT send the credentials (`Authorization: Negotiate someTicket`) if server did not provide a `WWW-Authenticate: Negotiate` header. 

Thus a trully DirectClient don't quite make sense here. As it is now, the 401 indicates to repeat the request with credentials, but does do a redirect, so it's in between direct and indirect, I guess?

---
in my case, I'd like to combine session/cookie storage with Kerberos, so would not need renegotiate with every request. so in brief what I'd like to have sounds like indirect client.

would this happen automatically if it implemented an indirect client?

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public KerberosClient(final Authenticator authenticator, final ProfileCreator<KerberosCredentials, KerberosProfile> profileCreator) {
+        setAuthenticator(authenticator);
+        setProfileCreator(profileCreator);
+    }
+
+    @Override
+    protected void clientInit(final WebContext context) {
+        setCredentialsExtractor(new KerberosExtractor(getName()));
```
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java]: We should have `defaultCredentialsExtractor(new KerberosExtractor(getName()));` to setup the `KerberosCredentialsExtractor` if none has already been set.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/profile/KerberosProfile.java (hunk 2 above)
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/profile/KerberosProfile.java]: OK

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/DummyAuthenticator.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package org.pac4j.kerberos.client.direct;
+
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.credentials.authenticator.Authenticator;
+import org.pac4j.core.exception.HttpAction;
+
+public class DummyAuthenticator implements Authenticator<Credentials> {
```
Comment (leleuj) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/DummyAuthenticator.java]: I generally don't create this kind of class but use lambdas for brevity in tests. That said, it's not a big deal.

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTests.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+    @Before
+    public void before() throws KrbException, IOException {
+        setupKerbyServer();
+    }
+
+    private void setupKerbyServer() throws KrbException, IOException {
+        kerbyServer = new SimpleKdcServer();
+        kerbyServer.setKdcHost("localhost");
+        kerbyServer.setKdcRealm("MYREALM.LT"); // FIXME
```
Comment (leleuj) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTests.java]: `FIXME`?

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientTests.java [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Test
+    public void testMissingProfileCreator() {
+        final KerberosClient kerberosClient = new KerberosClient(kerberosAuthenticator);
+        kerberosClient.setProfileCreator(null);
+        TestsHelper.initShouldFail(kerberosClient, "profileCreator cannot be null");
+    }
+
+    @Test
+    public void testBadAuthenticatorType() {
```
Comment (leleuj) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientTests.java]: Finally, this can be removed as well as the `DummyAuthenticator`.

File: pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java [hunk 14]
```diff
@@ -112,6 +112,19 @@ public static HttpAction unauthorizedDigest(final String message, final WebConte
...
 
     /**
+     * Build a response requesting to provide credentials via Kerberos/SPNEGO Negotiate mechanism.
+     *
+     * @param message message
+     * @param context context
+     * @return 401 Unauthorized with "WWW-Authenticate: Negotiate"
+     */
+    public static HttpAction unauthorizedNegotiate(final String message, final WebContext context) {
+        context.setResponseHeader(HttpConstants.AUTHENTICATE_HEADER, "Negotiate");
```
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java]: Maybe `Negotiate` would deserve a constant.

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTests.java [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * This is a test with real kerberos ticket validation:
+ * - spins up a mini/lightweight Kerberos server (using Apache Kerby)
+ * - generates the keytabs for service and client
+ * - generates a real service ticket, and passes it to mock context
+ * - checks that the ticket is correctly validated, and yields a correct client ID
+ *
+ * @author Vidmantas Zemleris, at Kensu.io
+ */
```
Comment (leleuj) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientKerbyTests.java]: `@since 2.1.0`

File: pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientTests.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+import javax.servlet.http.HttpServletResponse;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.util.Base64;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.*;
+
+
+public class KerberosClientTests implements TestsConstants {
```
Comment (leleuj) [inline pac4j-kerberos/src/test/java/org/pac4j/kerberos/client/direct/KerberosClientTests.java]: Javadoc...

**Comment by leleuj** [comment]
Overall, this is excellent. I made a few comments here and there, some things need to be fixed.

The tests are excellent as well.

The documentation is missing: the `README.md` must reference the Kerberos support and the website must be updated (new authenticator) via the files in the `documentation` directory. Please read: https://github.com/pac4j/pac4j/tree/master/documentation.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java (hunk 8 above)
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java]: is there any use of having `KerberosProfile#` prefix in profile ID (was in original code).
so I guess just delete the `L56-59` below.
earlier `subject/profileID = "garry"` after second assignment `subject/profileID=KerberosProfile#garry`.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java (hunk 10 above)
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java]: so to allow the default to be overridden...

**Comment by vidma** [comment]
so it seems to me I should split `KerberosClient` into two:
- **`IndirectKerberosClient`** - behaves like now, just extends `IndirectClient`. when used with `forceLogin`q, it seems to give this session persistence what I wanted.
- `DirectKerberosClient` - behaves like now, mostly.
  * not yet sure what it should do if no `Authenticate: Negotiate SomeTicket` header present:
     - `throw a HttpAction` in `retrieveCredentials` (like in `IndirectBasicAuthClient`) 
     - or simply`Extractor.extract() { return null }` like in `DirectBasicAuthClient`

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java (hunk 9 above)
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java]: Yes, you're completely right. The direct client is meant to take the credentials provided with the request and not break the flow if there are no credentials. So you should have no `throw HttpAction.unauthorizedNegotiate` in the direct client.

To keep the authenticated user profile in session, the indirect client seems to be the right choice, but the indirect client also redirects to the identity server for login (at the beginning of the flow). In the case of basic auth, the login URL is the callback URL (where credentials are expected) and in that case, if credentials are missing, a 401 error is returned.

My guess is that you should have something really really close to the basic auth support: `DirectBasicAuthClient`, `BasicAuthExtractor` and `IndirectBasicAuthClient`.

The `KerberosExtractor` should be shared among both clients, should not throw any `HttpAction` while the `KerberosIndirectClient` should handle that.

My only question is whether a direct client makes sense for Kerberos, but you know better...

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java (hunk 8 above)
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java]: No, when setting an identifier, you don't need to use the `#` separator, but when you request the typed identifier, you get `the class name + "#" + the identifier`.

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java (hunk 10 above)
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/client/direct/KerberosClient.java]: This won't work: `setCredentialsExtractor` sets the credentials extractor while `defaultCredentialsExtractor `sets the credentials extractor only if it is null (if is has not been set).

Example:

```java
KerberosClient client = new KerberosClient();
client.setAuthenticator(mySpecificAuthenticator);
```

`clientInit` is only called after the setter, so the authenticator must not be set to the default one, except if it is null.

**Comment by leleuj** [comment]
To sum up, you should have only one `KerberosExtractor` shared among the `DirectKerberosClient` and the `IndirectKerberosClient`. In the latter case, the indirect client should throw the 401 Negotiate response.

See the basic auth support, it seems to be the same...

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java (hunk 9 above)
Comment (vidma) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java]: yeah makes sense.
at least this time we gonna use an `IndirectClient`, so that's what I'll be focusing on.

> My only question is whether a direct client makes sense for Kerberos, but you know better...


good question for which I don't know a full answer. but if one is using Kerberos as a webservice/headless auth, the current `401` flow in direct client might make sense more than redirect with indirect client.
So to keep more flexibility for others, I'd just add an option in `DirectClient` -
 `FORCE_AUTH_IF_NO_CREDENTIALS` if true, when credentials are missing, a 401 should be returned.

File: pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java (hunk 14 above)
Comment (vidma) [inline pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java]: dunno, all other fns use string values too...

File: pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java (hunk 9 above)
Comment (leleuj) [inline pac4j-kerberos/src/main/java/org/pac4j/kerberos/credentials/extractor/KerberosExtractor.java]: Yes, good option

File: pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java (hunk 14 above)
Comment (leleuj) [inline pac4j-core/src/main/java/org/pac4j/core/exception/HttpAction.java]: OK. Keep things that way

**Comment by leleuj** [comment]
The build is broken due to PMD checks: `PMD Failure: org/pac4j/kerberos/credentials/authenticator/KerberosAuthenticator.java:8 Rule:UnusedImports Priority:4 Avoid unused imports such as 'org.pac4j.core.util.InitializableObject'.`

**Comment by vidma** [comment]
yeah, work still in progress.

so I think Direct/Indirect split should be +/- as it is. just I may add an optional `401` to force auth (as otherwise sometimes no ticket might be sent at all...), to DirectClient too. what do you think?

**Comment by leleuj** [comment]
"just I may add an optional 401 to force auth (as otherwise sometimes no ticket might be sent at all...), to DirectClient too. what do you think?": yes, but auth would not be forced by default

**Comment by vidma** [comment]
So currently this PR behaves very similarly to the BasicAuth.
As Direct and Indirect clients behave similarly in terms of Kerberos testing, I decided to test both in KerberosClientsKerbyTests.

still would like to retest with `play-pac4j` (getting some sort of jarhell; will check this on ~Monday), otherwise this is mostly finished, finally.

--

P.S. for `DirectClient` I decided **not to add** the `optional HTTP401 with WWW-Authenticate: Negotiate` yet. 
  * if someone has a clear use-case and needs, they can use it even without modifying `pac4j-kerberos` code, simply by defining their own `DirectClient`.

**Comment by leleuj** [comment]
You don't need to squash the commits, I'll use the appropriate Github option for the merge.

Yes, regarding the DirectClient, it's better not to make it blocking (although feasible).

Is this ready for merge or is there anything left?

**Comment by vidma** [comment]
yeah, this OK ready to merge now (changed 3 lines of docs recently).
tested OK with `play-pac4j` and real kerberos on Chrome, Firefox, and Safari.

**Comment by vidma** [comment]
@leleuj btw,  there's a related thing to disscuss for `play-pac4j` - we need `pac4j-kerberos` with play 2.3 or 2.4 (as play 2.5 don't support scala 2.10 anymore, and we still need scala 2.10).
  so looks like we need to backport `pac4j-kerberos` for an earlier version of `play-pac4j / pac4j`,  right (as master is for play 2.5)?

**Comment by vidma** [comment]
FYI: I backported pac4j-kerberos to `pac4j 1.7.x` (as required by the damn old play 2.3 we are still locked to) https://github.com/vidma/pac4j/pull/1

**Comment by leleuj** [comment]
Excellent work! Thanks. Merging.
